### DISCUSSION TITLE
Flexible initial pledge target lock

### DISCUSSION THREAD

**Discussion by AxCortes** [filecoin-project/FIPs#1044] [Enhancements - Technical]
## Background
The largest component of the initial pledge is the Consensus pledge. This contains a target lock variable of 0.3, which aims to keep around 30% of the circulating supply locked as initial pledge at all times. This design is insensitive to what is the actual demand for Filecoin. 

## Proposal
We propose a flexible target lock mechanism that would allow the total fraction of circulating supply to be locked at a given time to move above or below 30%, depending on demand for Filecoin usage.

## Current initial pledge target lock mechanism

The initial pledge has two components: *storage pledge* and *consensus pledge*.

The storage pledge is defined as an estimation of 20 days of expected block rewards. This component of the pledge is currently much smaller than the consensus pledge, so we will ignore it for the purposes of this discussion

The consensus pledge, $P$, for a given SP onboarding an amount $q$ of quality adjusted power is defined as

$$P=0.3S\frac{q}{{\rm Max}(Q,b)},$$

where $S$ is the circulating supply, $Q$ is the total network quality adjusted power, and $b$ is the baseline function

**The target lock is given by this numerical factor of 0.3** which is a fixed constant, and has important consequences on how much of the circulating is locked as initial pledge at any given time.

### Equilibrium total locked pledge

Suppose Filecoin reaches storage power onboarding equilibrium defined as: the rate at which new power is onboarded is the same as the rate at which old power expires.

If there is storage equilibrium *and* circulating supply stays constant, then exactly 30% of the circulating supply is locked.

In the real world, Filecoin is never exactly at equilibrium, and this means that total locked pledge will always hover near 30%, though not necessarily exactly at 30%.

## Proposed mechanism: explicitly value both storage and pledge

When a sector is onboarded, it adds value to the Filecoin network for two reasons: It adds to the total storage power of the network, and it increases the amount of pledge value locked. Block rewards are given to reward the SP's for this added value. 

It is, however, not explicitly stated how much of the reward is given because of the storage power added vs reward given because of the value locked through initial pledge. 

In other words, Filecoin is powered by a mixture of Proof of Useful Work and Proof of Stake, but it is not clear precisely how each of these components are valued.

The core of our proposal is: **Filecoin should give block rewards explicitly for PoUW and PoS components.** 

Suppose an SP is onboarding a sector that contains an amount of storage power, and an amount of initial pledge. We define

$$a={ \rm The \ amount \ of \ value\ the\ Filecoin\ network\ places\ on \ that\ sector's\ storage\ power,}$$

$$b={\rm  The\ amount\ of\ value\ the\ Filecoin\ network\ places\ on \ that\ sector's\ initial\ pledge.}$$

We can define the ratio,

$$\xi=\frac{b}{a},$$

and note that this parameter allows us to interpolate between a full Proof of stake network at $\xi\to\infty$ and a full Proof of Work at $\xi\to 0$.

### Multipliers for modified sectors

With these definitions of what is valuable to Filecoin, we can now ask questions such as **how much reward should Filecoin give to a sector with less (or more) pledge than required?**

A standard sector provides initial pledge which follows the current initial pledge formula. A modified sector includes the same storage power, but a different amount of pledge. The network can then give this modified sector a different amount of block rewards, based on a different multiplier. We define the multiplier $m$ as,

$$m=\frac{\rm Value\,to\,the\,network\,of\,modified\,sector}{\rm Value\,to\,the\,network\,of\,standard\,sector}$$

Suppose the modified sector has an initial pledge that is $x$ times the pledge of the standard sector. Then the multiplier can be calculated as,

$$ m=\frac{a+xb}{a+b}=\frac{1+x\xi}{1+\xi}$$

That is, if Filecoin knows how much value it assigns to storage power vs how much value it assigns to initial pledge, then it can accept modified sectors with different amounts of pledge, and it modifies their block rewards accordingly with a quality multiplier.

There remains the question of how Filecoin would choose their ratio $\xi$.

### Possible mechanism for fixing $\xi$

The mechanism we have presented works, as long as the network has a clear idea of how much it values storage power vs how much it values initial pledge. 

Here we present a possible solution, that works by still targetting the current 30%  locked supply, but in a much more flexible way, that allows SP's to lock more or less than that if they want to.

The idea is that the parameter $\xi$ can be adjusted depending on how much of the supply is actually locked. If there is less than 30% of supply locked, then $\xi$ would be larger, placing higher value on locking more token. If more than 30% of the supply is locked, then it is less important to lock more tokens, and $\xi$ can be lower.

We can define this more precisely. We define $L$ as the total amount of currently locked initial pledge. **In our proposal, the parameter $\xi$ can be adjusted based on the current value of $L$**

We can recreate the desired dynamic by fixing $\xi$ through a formula like,

$$\xi=f(0.3S-L),$$

where The function $f(x)$ has the following properties:
1) $f(x)$ is monotonically increasing with $x$.
2) $f(0)=1$.
3) $f(x)>0$ for all $x$.

We are purposely not completely specifying here what this function to be, to leave room for discussion and different ideas.

Note also that the target used here doesn't have to be the same 0.3 that is used in the current initial pledge formula.

**Comment by willscott** [comment]
This is a coherent design, I think the majority of contention will come from getting to alignment on the premise that we can get to an explicit division of a and b. I think there is a meaningful faction of the network that sees much of the long term value of filecoin in the useful work / storage market rather than the token staking. If one is coming in from that position, where $a/b > 10$, then the remainder of the mechanism presented is not particularly attractive.

I think this reinforces the need to get to a clearer alignment on how the network values these quantities and at what time scales. (one opportunity that may help to resolve the potential for conflict here is that we may find better alignment in separately valuing a vs b in the 1 year timeframe versus a longer time frame)

**Reply by AxCortes** [reply]
I think Filecoin currently does place value on both the storage power, and the initial pledge, at least that's what the 0.3 target reflects.

So my proposal is aiming to introduce more flexibility, while keeping about the same value for pledge vs storage as we have now. So it introduces the flexibility while keeping the same 0.3 parameter.

I think if people want to argue that we are overvaluing pledge, this should be a separate discussion, on whether we should change the target from 0.3 or not.

**Comment by dd45e640b42e6da7da96faee3996ef7c** [comment]
i propose to set the target to 0. its the most capital efficient way to provide both market independent consensus security and "storage". vote by raw byte power to change it.

without a hard, on chain, binding voting process i am against implementing this. it will intro unnecessary friction.

**Reply by AxCortes** [reply]
I think that is a valid, but separate proposal. In this proposal, I am keeping the current 0.3 target, but adding more flexibility to it, so it won't result in 30% always being locked regardless of demand.

One could also argue that the target should be more or less than 0.3, but I think that is a different discussion.

**Reply by dd45e640b42e6da7da96faee3996ef7c** [reply]
the title threw me off tbh

> Flexible initial pledge target lock

thats not what this proposal is discussing i think. 

> One could also argue that the target should be more or less than 0.3, but I think that is a different discussion.

**Comment by Fatman13** [comment]
> People who have always operated without skin in the game (or without their skin in the right game) seek the complicated and centralized, and avoid the simple like the plague. Practitioners, on the other hand, have opposite instincts, looking for the simplest heuristics. Some rules:

> *People who are bred, selected, and compensated to find complicated solutions do not have an incentive to implement simplified ones.*

> Excerpt From: Nassim Nicholas Taleb. “Skin in the Game.” Apple Books.

**Reply by AxCortes** [reply]
Yes, this proposal introduces a trade off:

The current mechanism is Simple, but very rigid, and leads to the situation I am discussing, that 30% is locked at all times.

My mechanism proposal allows us to move from fixed 30%, at the cost of the mechanism being more complicated, this is true.
If there is an alternate solution that is simpler than my proposal, while solving the same problem, that would be even better.

Then we have the discussion to figure out where the community wants to land on this tradeoff. 

Whether I have skin in a game or not, should be irrelevant to discussing the merits or shortcomings of the proposal

**Comment by anorth** [comment]
I like a lot of where this proposal is coming from. However, I think it is too bound by attempting to hew to the existing Filecoin specification. The 30% target lock in the Filecoin spec is one of the most problematic aspects of the cryptoeconomic setup, as it effectively limits the locking of tokens at a level far below what we might expect in comparable networks.

I would like to explore variations of this proposal that do not encode an explicit target, but adjust the ratio of rewards paid to storage capacity versus pledge dynamically according to the flows and availability. I have _not worked this through, it's just inspiration, probably naive_, but for example:
- The network tracks a current level of total pledge (P), total storage capacity (S) and total money supply (T, meaning minted less burned).
- The network maintains one new number X in [0, 1] which is the fraction of block rewards paid to pledge vs storage. It can be initially set from network facts of storage power and pledge (somewhere a bit <90%, I would guess).
- Upon a new unit P' of pledge being added, X is increased linearly by `P' / (T-P)`, i.e, along a straight line such that X would reach 1.0 if the entire money supply were pledged (note this point will never be reached).
- Upon a new unit S' of capacity being added, X is decreased by `X/S`, i.e. along a hyperbola such that X would approach 0.0 if infinite capacity were added.

The mechanism targets _as much as possible_ of both storage and pledge, there is no 30%. But note critically that the total pledge is bounded by the total token supply, while total storage is unbounded. From a hypothetical extreme of some storage capacity and little pledge, the mechanism would strongly favour adding more pledge, but this would only continue until the yield on that pledge diluted to some risk-compatible amount (see other networks). From the other hypothetical of 100% of supply pledged, storage is incentivised, but unbounded from the network POV (in reality, by the ROI). From that point, unlimited of storage can be added and keep moving X toward storage, because _there are no more tokens to pledge to move it the other way_. So long as hardware efficiency keeps improving, the incentive for more storage is unbounded. Of course, the 100% pledge could never be reached, but I expect a similar dynamic would apply at whatever the saturating pledge level is.

I would expect that the equilibrium level of pledge is well above 30%, so this would contribute a great near-term deflationary effect. Initially this would reduce the rewards due to raw storage, but once the pledge saturates, the dynamics would turn back to incentivising storage growth. I haven't done any modelling, but my intuition is that the network would tend to a state with much more locking, but still a similar or greater share of reward paid to storage (increasing over the long term)

Initially I would say pledge is still attached to sectors, as in your proposal, but this per-sector attachment becomes a bit dubious (in both). Simpler, a miner could have some sectors and a pool of pledge separately. (Ref also #972, #1036)

**Reply by AxCortes** [reply]
Thanks for the suggestions, good to see some alternate designs!

I kept the 30% mostly from the philosophy of not changing many things at once. I just aimed to introduce some flexibility to the target lock, as the title said, while not necessarily explicitly moving it from 30%. For sure there is conversation to be had about whether 30% is wrong.

I like the principle of trying to design a mechanism that is "parameter free", as long as it doesn't become too complex either, which would be another trade off. I have to spend some more time understanding your proposal, if it is a feasible way to acheive this, and what are the consequences.

So you are saying $X=1/\xi$ in my proposal, I think?  You gave me some kind of rules on how to update X, but then I don't understand what the initial setting is, or better yet, what the formula for $X$ in terms of $S,T,P$ is. For instance it could be that the parameter 0.3 is hidden in there, if we say something like "let's initialize it to the conditions we have now" which is that 30% of supply being locked is normal starting point. 

So I guess first step would be actually writing down what a formula for $X$ here would look like, and then we can ask what are the consequences, and compare with my version?

**Reply by anorth** [reply]
Yes I am thinking "let's initialize it to the conditions we have now", so the 30% is in there as part of initial conditions, but is not in the update rule and should reduce in influence over time.

I don't think one can write X in terms of S, T, P because the current value depends on the history of the network evolution, not just those values today. But it is very easy to just read the value of X out of network state. Note that really X (today) is per sector, depending on the pledge:power ratio, and would vary between miners with different histories, but I propose just aggregating it for the network here.

**Comment by sviperj** [comment]
Good Idea. Support.
